main_test.go

```diff
@@ -281,10 +330,498 @@ func TestControllerConfigmapUpdate(t *testing.T) {
+						AZ:      "hashring0",
+					},
+					{
+						Address: "hashring0-1.h0.namespace.svc.cluster.local:10901",
+						AZ:      "hashring0",
+					},
+					{
+						Address: "hashring0-2.h0.namespace.svc.cluster.local:10901",
+						AZ:      "hashring0",
+					},
+				},
+			}},
+		},
+		{
+			name: "OneHashringManyStatefulSets",
+			hashrings: []receive.HashringConfig{{
+				Hashring: "hashring0",
+				Tenants:  []string{"foo", "bar"},
+			}, {
+				Hashring: "hashring1",
+			}},
+			statefulsets: []*appsv1.StatefulSet{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "hashring0",
+						Labels: map[string]string{
+							"a":              "b",
+							hashringLabelKey: "hashring0",
+						},
+					},
+					Spec: appsv1.StatefulSetSpec{
+						Replicas:    intPointer(3),
+						ServiceName: "h0",
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "hashring1",
+						Labels: map[string]string{
+							"a":              "b",
+							hashringLabelKey: "hashring1",
+						},
+					},
+					Spec: appsv1.StatefulSetSpec{
+						Replicas:    intPointer(2),
+						ServiceName: "h1",
+					},
+				},
+			},
+			clusterDomain: "cluster.local",
+			expected: []receive.HashringConfig{{
+				Hashring: "hashring0",
+				Tenants:  []string{"foo", "bar"},
+				Endpoints: []receive.Endpoint{
+					{
+						Address: "hashring0-0.h0.namespace.svc.cluster.local:10901",
+						AZ:      "hashring0",
+					},
+					{
+						Address: "hashring0-1.h0.namespace.svc.cluster.local:10901",
+						AZ:      "hashring0",
+					},
+					{
+						Address: "hashring0-2.h0.namespace.svc.cluster.local:10901",
+						AZ:      "hashring0",
+					},
+				},
+			}, {
+				Hashring: "hashring1",
+				Endpoints: []receive.Endpoint{
+					{
+						Address: "hashring1-0.h1.namespace.svc.cluster.local:10901",
+						AZ:      "hashring1",
+					},
+					{
+						Address: "hashring1-1.h1.namespace.svc.cluster.local:10901",
+						AZ:      "hashring1",
+					},
+				},
+			}},
+		},
+		{
+			name: "OneHashringLabelKeyManyStatefulSets",
+			hashrings: []receive.HashringConfig{{
+				Hashring: "hashring0",
+				Tenants:  []string{"foo", "bar"},
+			}},
+			statefulsets: []*appsv1.StatefulSet{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "hashring0",
+						Labels: map[string]string{
+							"a":              "b",
+							hashringLabelKey: "hashring0",
+						},
+					},
+					Spec: appsv1.StatefulSetSpec{
+						Replicas:    intPointer(3),
+						ServiceName: "h0",
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "hashring1",
+						Labels: map[string]string{
+							"a":              "b",
+							hashringLabelKey: "hashring0",
+						},
+					},
+					Spec: appsv1.StatefulSetSpec{
+						Replicas:    intPointer(2),
+						ServiceName: "h0",
+					},
+				},
+			},
+			clusterDomain: "cluster.local",
+			expected: []receive.HashringConfig{
+				{
+					Hashring: "hashring0",
+					Tenants:  []string{"foo", "bar"},
```

Comment:
how the tenants populate?

Reply:
tenants will not change throughout the populate method, once you provided tenant info in intial configmap, it will stay there forever